### PR TITLE
feat: let the Primary's core use the Consensus recovery

### PR DESCRIPTION
**Pull request by huitseeker**
This is a superset of MystenLabs/narwhal#814:
- this front-loads the recovery of the consensus, creating its state before actually spawning the task,
- then modifies the recovery of the primary to use the result of consensus recovery,

Note that we do not provide a recovery to the primary in case we are running with an external consensus. In this case, we simply follow the logic outlined in MystenLabs/narwhal#814.

Opening for review for comments, this requires more testing.

Part of MystenLabs/sui#5324

### CODE REVIEW

**Review by akichidis — COMMENTED** [review]
I really like the approach you took on this @huitseeker ! Agreed we need more testing. Left a couple of comments, let me know your thoughts.

File: primary/src/core.rs [hunk 1]
```diff
@@ -134,11 +136,92 @@ impl Core {
...
+    pub async fn recover(mut self, opt_consensus_state: Option<ConsensusState>) -> Self {
+        info!("Start recovering process. Will not process any message until finished.");
+
+        let now = Instant::now();
+
+        // Fetch the certificates that are higher or equal than the last_commit_round
+        // We need that as an indication to filter the certificates after a certain
+        // point that we are confident that we have reached in the past. Then we can
+        // detect the last.
+        let certificates_after_gc_round = if let Some(consensus_state) = opt_consensus_state {
```
[the remark this answers is not in the document]
Comment (huitseeker) [reply primary/src/core.rs:156]: The computation of the `last_round` requires a scan of the DB, which is done in the `ConsensusState` (unavoidably):
https://github.com/MystenLabs/narwhal/pull/102/files#diff-33a8786eccaee3bdd04de13118a2341ad3f7ecad3fe329b1b2548555fbf0cad4
Here I'm simply passing that result along to the `Core`, which avoids recomputing its initial value.

[the remark this answers is not in the document]
Comment (huitseeker) [reply primary/src/core.rs:156]: With that said, once you further introduce the ability to compute the `last_round` from the Certificate store, i agree we can scrap the part injecting the Consensus reconstruction.

**Comment by huitseeker** [comment]
The simplest way to bring this to a close (w/o spurious code changes) is that this will need to be refactored on top of (and is blocked by) MystenLabs/narwhal#825.

**Review by lanvidr — COMMENTED** [review]
what further testing do you have in mind for this?

**Comment by akichidis** [comment]
Closing the PR as newer work has been done under Narwhal's new home https://github.com/MystenLabs/sui

**Pull request closed without merging**